src/adagrams.js

```diff
@@ -1,15 +1,142 @@
+//WAVE 1
 export const drawLetters = () => {
-  // Implement this method for wave 1
-};
+  const LETTER_POOL = {
+    A: 9,
+    B: 2,
+    C: 2,
+    D: 4,
+    E: 12,
+    F: 2,
+    G: 3,
+    H: 2,
+    I: 9,
+    J: 1,
+    K: 1,
+    L: 4,
+    M: 2,
+    N: 6,
+    O: 8,
+    P: 2,
+    Q: 1,
+    R: 6,
+    S: 4,
+    T: 6,
+    U: 4,
+    V: 2,
+    W: 2,
+    X: 1,
+    Y: 2,
+    Z: 1,
+  };
+
+  let letterBank = [];
+  let letterCopy = JSON.parse(JSON.stringify(LETTER_POOL));
+  let keyList = Object.keys(letterCopy);
 
+  while (letterBank.length < 10) {
+    let randomLetter = keyList[Math.floor(Math.random() * keyList.length)];
+    if (letterCopy[randomLetter] > 0) {
+      letterBank.push(randomLetter);
+      letterCopy[randomLetter]--;
+    }
+  }
+  return letterBank;
+};
+//WAVE 2
 export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
+  let letters = input.toUpperCase();
+
+  //for-of loop iterates over the values in an iterable object, such as an array.
+  for (const letter of letters) {
+    if (!lettersInHand.includes(letter)) {
+      return false;
+    }
+  }
+
+  const letterCounts = {};
+  for (const letter of letters) {
+    if (!letterCounts[letter]) {
+      letterCounts[letter] = 1;
+    } else {
+      letterCounts[letter]++;
+    }
+  }
+
+  for (const [letter, count] of Object.entries(letterCounts)) {
+    if (count > lettersInHand.filter((l) => l === letter).length) {
+      return false;
+    }
+  }
+  return true;
 };
 
+//WAVE 3
 export const scoreWord = (word) => {
-  // Implement this method for wave 3
+  const SCORE_CHART = {
+    A: 1,
+    B: 3,
+    C: 3,
+    D: 2,
+    E: 1,
+    F: 4,
+    G: 2,
+    H: 4,
+    I: 1,
+    J: 8,
+    K: 5,
+    L: 1,
+    M: 3,
+    N: 1,
+    O: 1,
+    P: 3,
+    Q: 10,
+    R: 1,
+    S: 1,
+    T: 1,
+    U: 1,
+    V: 4,
+    W: 4,
+    X: 8,
+    Y: 4,
+    Z: 10,
+  };
+
+  let score = 0;
+  const letters = word.toUpperCase();
+  for (const letter of letters) {
+    score += SCORE_CHART[letter] || 0;
+  }
+
+  if (letters.length >= 7) {
+    score += 8;
+  }
+  return score;
 };
 
+//WAVE 4
 export const highestScoreFrom = (words) => {
-  // Implement this method for wave 4
+  let winningWords = {};
+  let winningScore = 0;
+
+  for (const word of words) {
+    const score = scoreWord(word);
+
+    if (score > winningScore) {
+      winningScore = score;
+      winningWords["score"] = score;
+      winningWords["word"] = word;
```

Comment:
This works but you could simplify this by creating one variable to hold the winning data. Using two variables takes up more memory and most primitives (booleans, numbers, null, undefined, etc.) are constant (O(1)) space. This is because, no matter if true or false, or 1 or 1,000, the amount of space consumed stays constant. Reference types (arrays and objects) — Reference types require O(n) space as well. ’n’ being the length for an array, or in the case of an object, ’n’ being the amount of keys. Reference types are O(n) because the more elements in an array, or the more keys in an object, the more space they consume. The `winningWords` object does have a space complexity of 0(n), while the `winningScore` number has a space complexity of 0(1), but since that's less than the `winningWords` variable, we drop it. You are also using a score `variable` which holds the result of `scoreWord`.